diff --git a/utf8/cweav-utf8.ch b/utf8/cweav-utf8.ch
index dbe135c..eb2a7f3 100644
--- a/utf8/cweav-utf8.ch
+++ b/utf8/cweav-utf8.ch
@@ -50,9 +50,11 @@ int mosntowcs(char *mbs, int len, wchar_t *s)
   int n = 0;
   int l = 0;
   while (l<len) {
-    l+=motowc(mbs+l, s, len-l);
-    n++;
-    if (s!=NULL) s++;
+    int x = mbtowc(s, mbs+l, MB_CUR_MAX);
+    if (x == -1) l+=1;
+    else l+=x;
+    if (x != -1) n++;
+    if (x != -1) if (s!=NULL) s++;
   }
   if (s!=NULL) *s=L'\0';
   return n;
